File: backend/internal/faasproto/frame.go
// Package faasproto is the M4 wire protocol between the FaaS supervisor (Go) and the
// Bun worker: length-prefixed, binary-safe frames carrying VALUES only (never the
// secrets key or a DB handle, D18.7/D24.4). The response image is a FIXED-SIZE raw
// RGB block — ReadFrame enforces a hard ceiling BEFORE any allocation and DecodeResponse
// rejects any image byte count other than the exact 400x300 raw size, so a malicious
// worker can neither over-declare a length to OOM the key+DB supervisor nor smuggle a
// compressed payload it would have to decode (fix-4).
//
// Proto v2 (A27 W3 / K7) splits the single frame ceiling into two DIRECTION-specific caps.
// The sup→worker REQUEST may now carry a multi-MB source image (Request.Input, base64), so
// WriteFrame (which the trusted supervisor uses to WRITE requests) is bounded by the larger
// MaxRequestFrame. The worker→sup RESPONSE stays tight: ReadFrame (which the trusted supervisor
// uses to READ the untrusted worker's response) keeps MaxFrame, so a malicious worker still
// cannot over-declare a multi-MB response length to OOM the key+DB supervisor (T4). One constant
// can never be both tight (response) and multi-MB (request) — hence two.
package faasproto

import (
	"encoding/binary"
	"fmt"
	"io"
)

// Frame kinds (the u8 that leads every message payload).
const (
	KindRenderRequest  = 0x01 // sup -> worker
	KindRenderResponse = 0x02 // worker -> sup
	KindCancel         = 0x03 // sup -> worker   payload = JSON {id}
	KindReady          = 0x04 // worker -> sup   payload = JSON {slot, idle}
)

const (
	// RawFrameSize is the fixed raw RGB frame the worker returns: 400*300*3. The panel
	// geometry is fixed (bwry.Width/Height); a mismatch here is a protocol violation.
	RawFrameSize = 400 * 300 * 3 // 360000
	// MetaMax caps the response meta JSON (small: dither/hint/log/err).
	MetaMax = 16 << 10 // 16 KiB
	// MaxFrame bounds the worker→sup RESPONSE direction before allocation: kind(1) + metaLen(4)
	// + meta + raw image. ReadFrame (the trusted supervisor reading the untrusted worker) enforces
	// it, so a malicious worker cannot over-declare a giant length to OOM the key+DB supervisor
	// (T4). Kept tight — it does NOT cover the request's source image (that is MaxRequestFrame).
	MaxFrame = 1 + 4 + MetaMax + RawFrameSize
	// MaxRequestImageBytes caps the source image the built-in __playlist source receives in a
	// render request (Request.Input). Multi-MB source blobs (§4.3), aligned with the imgstore
	// ingest ceiling (imgstore.MaxImagePixels is a pixel cap; this is the request-frame byte cap).
	MaxRequestImageBytes = 12 << 20 // 12 MiB
	// MaxRequestFrame bounds the sup→worker REQUEST direction: kind(1) + the request JSON envelope
	// (source + secret values + ctx + limits, MetaMax slack) + the base64-encoded source image.
	// SEPARATE from MaxFrame (K7): only the trusted supervisor produces this direction, so a larger
	// ceiling here does NOT weaken the T4 response wall. base64 inflates the raw image by 4/3.
	MaxRequestFrame = 1 + MetaMax + (MaxRequestImageBytes/3+1)*4
)

// WriteFrame writes one length-prefixed message: u32be(total) | kind | payload, where
// total = 1 + len(payload). In Go this is the supervisor→worker REQUEST writer (driveWorker),
// so the ceiling is MaxRequestFrame — the request may carry a multi-MB source image (K7). The
// response direction is written by the TS worker (encodeFrame, MaxFrame) and read here by
// ReadFrame (MaxFrame), which keeps the T4 wall to the untrusted worker.
func WriteFrame(w io.Writer, kind byte, payload []byte) error {
	total := 1 + len(payload)
	if total > MaxRequestFrame {
		return fmt.Errorf("faasproto: frame len %d > max %d", total, MaxRequestFrame)
	}
	buf := make([]byte, 4+total)
	binary.BigEndian.PutUint32(buf[:4], uint32(total))
	buf[4] = kind
	copy(buf[5:], payload)
	_, err := w.Write(buf)
	return err
}

// ReadFrame reads one message. It reads the 4-byte length prefix, REJECTS a declared
// length > MaxFrame (or < 1) BEFORE allocating the payload buffer (the OOM ceiling,
// T4), then reads exactly that many bytes and splits off the kind.
func ReadFrame(r io.Reader) (kind byte, payload []byte, err error) {
	var lenb [4]byte
	if _, err = io.ReadFull(r, lenb[:]); err != nil {
		return 0, nil, err
	}
	total := binary.BigEndian.Uint32(lenb[:])
	if total < 1 {
		return 0, nil, fmt.Errorf("faasproto: frame len %d < 1", total)
	}
	if total > MaxFrame {
		return 0, nil, fmt.Errorf("faasproto: frame len %d > max %d (rejected before alloc)", total, MaxFrame)
	}
	buf := make([]byte, total)
	if _, err = io.ReadFull(r, buf); err != nil {
		return 0, nil, err
	}
	return buf[0], buf[1:], nil
}

File: backend/worker/frame.ts
// M4 wire protocol (TS side) — mirrors internal/faasproto byte-for-byte:
//   message := u32be total_len | u8 kind | payload[total_len-1]
//   render-response payload := u32be metaLen | meta JSON | raw image bytes
// The end-to-end parity (Go supervisor <-> this worker over the UDS) is exercised in W4.

export const KIND_RENDER_REQUEST = 0x01;
export const KIND_RENDER_RESPONSE = 0x02;
export const KIND_CANCEL = 0x03;
export const KIND_READY = 0x04;

export const RAW_FRAME_SIZE = 400 * 300 * 3; // 360000
export const META_MAX = 16 << 10; // 16384
export const MAX_FRAME = 1 + 4 + META_MAX + RAW_FRAME_SIZE;

function beU32(b: Uint8Array, off: number): number {
  // multiply the high byte to stay positive (lengths here are far below 2^31 anyway)
  return b[off] * 0x1000000 + (b[off + 1] << 16) + (b[off + 2] << 8) + b[off + 3];
}

function putU32(b: Uint8Array, off: number, v: number): void {
  b[off] = (v >>> 24) & 0xff;
  b[off + 1] = (v >>> 16) & 0xff;
  b[off + 2] = (v >>> 8) & 0xff;
  b[off + 3] = v & 0xff;
}

// encodeFrame builds one length-prefixed message: u32be(1+payload) | kind | payload.
export function encodeFrame(kind: number, payload: Uint8Array): Uint8Array {
  const total = 1 + payload.length;
  if (total > MAX_FRAME) throw new Error(`faasproto: frame len ${total} > max ${MAX_FRAME}`);
  const out = new Uint8Array(4 + total);
  putU32(out, 0, total);
  out[4] = kind;
  out.set(payload, 5);
  return out;
}

export interface Frame {
  kind: number;
  payload: Uint8Array;
}

// FrameReader accumulates stream chunks and yields complete frames. It rejects a declared
// length > MAX_FRAME (or < 1) — the same ceiling faasproto.ReadFrame enforces.
export class FrameReader {
  private buf = new Uint8Array(0);

  feed(chunk: Uint8Array): Frame[] {
    const merged = new Uint8Array(this.buf.length + chunk.length);
    merged.set(this.buf, 0);
    merged.set(chunk, this.buf.length);
    this.buf = merged;

    const out: Frame[] = [];
    for (;;) {
      if (this.buf.length < 4) break;
      const total = beU32(this.buf, 0);
      if (total < 1) throw new Error(`faasproto: frame len ${total} < 1`);
      if (total > MAX_FRAME) throw new Error(`faasproto: frame len ${total} > max ${MAX_FRAME}`);
      if (this.buf.length < 4 + total) break;
      out.push({ kind: this.buf[4], payload: this.buf.slice(5, 4 + total) });
      this.buf = this.buf.slice(4 + total);
    }
    return out;
  }
}

export interface RenderErr {
  kind: string; // throw|timeout|egress|offsize|oom
  msg: string;
}

export interface ResponseMeta {
  v: number;
  id: string;
  ok: boolean;
  w: number;
  h: number;
  dither?: string;
  next_wake_hint?: number;
  log?: { lvl: string; msg: string }[];
  err: RenderErr | null;
}

// encodeResponse builds the render-response payload: u32be metaLen | meta JSON | raw. An OK
// response MUST carry exactly RAW_FRAME_SIZE image bytes; an error response MUST carry none
// (the fixed-size invariant faasproto.DecodeResponse enforces on the Go side).
export function encodeResponse(meta: ResponseMeta, raw: Uint8Array | null): Uint8Array {
  const rawLen = raw ? raw.length : 0;
  if (!meta.err && rawLen !== RAW_FRAME_SIZE) {
    throw new Error(`faasproto: ok response image ${rawLen} != ${RAW_FRAME_SIZE}`);
  }
  if (meta.err && rawLen !== 0) {
    throw new Error(`faasproto: err response must carry no image (${rawLen})`);
  }
  const mj = new TextEncoder().encode(JSON.stringify(meta));
  if (mj.length > META_MAX) throw new Error(`faasproto: meta ${mj.length} > max ${META_MAX}`);
  const out = new Uint8Array(4 + mj.length + rawLen);
  putU32(out, 0, mj.length);
  out.set(mj, 4);
  if (raw) out.set(raw, 4 + mj.length);
  return out;
}

export interface RenderRequest {
  v: number;
  id: string;
  fn: { id: number; ver: number; source: string };
  ctx: { serial: string; channel: string; trigger: { type: string; payload?: unknown }; now: string };
  force: boolean;
  dither_default: string;
  secrets: Record<string, string>;
  egress_allow: string[];
  egress_cred: string;
  limits: { timeout_ms: number; mem_mb: number };
}

export function decodeRequest(payload: Uint8Array): RenderRequest {
  return JSON.parse(new TextDecoder().decode(payload)) as RenderRequest;
}
